Add Working Groups Process

caniszczyk opened the pull request:
I've had a few set of folks come to me and want to propose a WG to CNCF. We've always had a process in place but never written down so this is an attempt to do that.

Signed-off-by: Chris Aniszczyk <caniszczyk@gmail.com>

caniszczyk commented:
RFC @cncf/toc, this essentially writes down what we have been doing. Just like a project proposal, any WG proposed has to have a proposal + TOC presentation + sponsor/vote.

I'm doing this now as there's a few folks in the community that want to propose new WGs so I'd like for us to solidify a simple process based on what we already do.

leecalcote commented on workinggroups/README.md:
```diff
@@ -0,0 +1,11 @@
+
+If you would like to submit a working group proposal, please submit a pull request to the working groups folder. As an example, you can see the other working group proposals here: https://github.com/cncf/toc/tree/master/workinggroups
+
+You will also have to present to the CNCF TOC and wider community before your WG proposal will be voted upon by the TOC and community. You can request a presentation by filing an issue here: https://github.com/cncf/toc/issues
```
## Expectations

Working Group representative(s) (either the TOC Sponsor or other representative) will provide updates to the TOC with some regularity (either as requested by the TOC or when meaningful WG milestone has been met).

wking replied:
This is asking for both a pull-request (with the proposal) and an issue (for the presentation request).  Should those be rolled into a single pull-requests which carries that proposal and acts as a presentation request?  If not, what's the benefit to keeping them separate?

bgrant0607 replied:
For projects, the process has been to first request a presentation opportunity, and then, if invited to do so by the TOC, submit an official proposal. Most projects that have presented have not been asked to submit proposals.

leecalcote commented on workinggroups/networking.md:
```diff
@@ -0,0 +1,53 @@
+
+## Objective
+
+Explore cloud native networking technology and concepts around the container networking interface (CNI).
```
## Objective

Explore cloud native networking technology, concepts around the container networking interface (connectivity), network services, common design patterns and best practices.

bgrant0607 replied:
"Explore" seems like an activity rather than an objective. Is there an objective at this point?

leecalcote commented on workinggroups/serverless.md:
```diff
@@ -0,0 +1,39 @@
+
+* Produce a whitepaper
+* Produce a serverless landscape
+* Explore specifications for serverless to propose to the CNCF
```
Explore specifications for serverless (e.g. events) to propose to the CNCF.

bgrant0607 replied:
s/Explore/Develop/

"Explore" isn't a goal.

mattfarina replied:
Is the WG evaluating existing specs or creating their own? Could "evaluate" be used instead of "explore"?

duglin commented on workinggroups/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Introduction
+
+The purpose of working groups are to study and report on a particular question and make recommendations based on its findings. The end result of a working group may be a new project proposal, landscape, whitepaper or even a report detailing their findings. The intention of working groups is not to host a full project or specification. Working Groups can be formed at any time but must be sponsored ...
```
"The intention of working groups is not to host a full project or specification."  What does this mean for the CloudEvents stuff?

caniszczyk commented:
CloudEvents should be proposed to the TOC eventually.

On Tue, Apr 3, 2018 at 11:49 PM, Doug Davis <notifications@github.com>
wrote:

> *@duglin* commented on this pull request.
> ------------------------------
>
> In workinggroups/README.md
> <https://github.com/cncf/toc/pull/106#discussion_r178872807>:
>
> > @@ -0,0 +1,11 @@
> +# CNCF Working Groups
> +
> +## Introduction
> +
> +The purpose of working groups are to study and report on a particular question and make recommendations based on its findings. The end result of a working group may be a new project proposal, landscape, whitepaper or even a report detailing their findings. The intention of working groups is not to host a full project or specification. Working Groups can be formed at any time but must be sponsored by a TOC member and voted with a super majority vote by the CNCF TOC. The TOC can also shut down a working group with a super majority vote.
>
> "The intention of working groups is not to host a full project or
> specification." What does this mean for the CloudEvents stuff?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cncf/toc/pull/106#pullrequestreview-109007549>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD5IbiVtgN3aR7Nckn1FnGuun0_uhcdks5tk5ojgaJpZM4TENpW>
> .
>



-- 
Cheers,

Chris Aniszczyk
http://aniszczyk.org
+1 512 961 6719

caniszczyk commented:
cc: @mattfarina

monadic commented:
cc @ultrasaurus would love your thoughts too

mattfarina commented on workinggroups/ci.md:
```diff
@@ -0,0 +1,34 @@
+
+## Non Goals
+
+* Run CI for CNCF projects
```
Does this non-goal conflict with the first goal of:

> We believe that it would be good for us to provide CI services to projects who need or want to use them

mattfarina commented on workinggroups/networking.md:
```diff
@@ -0,0 +1,53 @@
+* Not going to make a one network standard for all
+* Not going to focus on individual projects per service but rather projects that model network services and patterns not going to be prescriptive but more reference guidelines and patterns
+
+## Interested Parties
```
On the other WG Proposals this is called _Initial Interested Parties_. Is this meant to be an initial list only or a living list that changes over time?

mattfarina commented on workinggroups/ci.md:
```diff
@@ -0,0 +1,34 @@
+* We believe that it would be good for us to provide CI services to projects who need or want to use them
+* We need to understanding what, if any, SLA we can promise projects for this system
+* We need to scope what features this system will provide; there is some concern around trying to promise testing the full cross-product of integration with all of the different CNCF projects
+* We want to come away with a recommendation for staffing to support building out this initiative given project needs and desired SLA
```
Are working groups expected to be long lived (e.g., still doing work in 20 years), working to a complete a task, or may there be cases of both? I ask because some of these, such as storage and networking I could see some day coming to a close while the CI one could go on indefinitely if the first bullet is true. It might be useful in the readme to describe expectations around and for the two types of WGs if they diverge on this point. If some are expected to close it might be useful to have a definition of done in the proposal.

bgrant0607 replied:
I agree. The CI WG is different kind of entity from the others. It is producing code for the CNCF that isn't a project unto itself, but is in service of the CNCF's projects.

ultrasaurus replied:
I agree that the CI WG proposal is a bit different from what all the WGs are currently doing.  If a group of people are needed to support CNCF, what about making a CNCF CI team -- or some other word that we can use to convey a very different kind of activity.

mattfarina commented on workinggroups/serverless.md:
```diff
@@ -0,0 +1,39 @@
+* Produce a whitepaper
+* Produce a serverless landscape
+* Explore specifications for serverless to propose to the CNCF
+* Bring recommendations to the TOC on serverless projects in CNCF
```
This may be a bit of a sidebar but, is the CNCF looking to bring projects of certain types into its fold? Or, would it rather projects approach them? I'm curious on this point.

mattfarina commented on workinggroups/serverless.md:
```diff
@@ -0,0 +1,39 @@
+* Chris Munns (AWS)
+* Ken Owens (Mastercard)
+* Mark Peek (VMWare)
+* Cathy Zhang (Huawei)
```
Sometimes companies are listed and sometimes GitHub handles are listed (e.g., on the networking WG proposal). Should this be consistent between proposals? If we are going to pick one I would like to see the GitHub handle but I would prefer to record both. It's nice to see the company at a glance and have a quick way to find a GitHub id for more info.

nerdyyatrice replied:
I wonder what qualified a company to be in this list?

dankohn replied:
That was the original group of volunteers. This is the process for companies to be represented in CloudEvents: https://github.com/cloudevents/spec/blob/master/GOVERNANCE.md#voting

bgrant0607 commented on workinggroups/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Introduction
+
+The purpose of working groups are to study and report on a particular question and make recommendations based on its findings. The end result of a working group may be a new project proposal, landscape, whitepaper or even a report detailing their findings. The intention of working groups is not to host a full project or specification. Working Groups can be formed at any time but must be sponsored ...
```
It is very important that every WG be created with a charter that explicitly states the mission of the WG, what its concrete deliverables are expected to be, and what it's exit criteria are. WGs should exist forever. They should address specific issues or questions.

bgrant0607 replied:
And if they publish information under the CNCF, such as landscapes or whitepapers, I think it makes sense to at least send them to the TOC for review, possibly with an explicit time bound for feedback ("RFC" period).

quinton-hoole replied:
Presumably you mean "WG's should NOT exist forever"?

mattfarina replied:
> And if they publish information under the CNCF, such as landscapes or whitepapers, I think it makes sense to at least send them to the TOC for review, possibly with an explicit time bound for feedback ("RFC" period).

Should it be implicit or explicit? For example, is lazy-concensus ok?

bgrant0607 commented on workinggroups/networking.md:
```diff
@@ -0,0 +1,53 @@
+
+## Goals and Expected Outcomes
+
+* Recommend CNI be adopted as initial network interface specification focused on connectivity and portability as an official CNCF project.
```
I assume this is done. Otherwise, it should be under the CNI project rather than in a WG.

bgrant0607 commented on workinggroups/networking.md:
```diff
@@ -0,0 +1,53 @@
+
+Explore cloud native networking technology and concepts around the container networking interface (CNI).
+
+## Goals and Expected Outcomes
```
Does this WG still need to exist, or can it declare victory and dissolve?

mattfarina replied:
This WG exists today. Since this PR is about adding process and listing the current WGs should the discussion on dissolving the WG be moved to a meeting or mailing list?

bgrant0607 commented on workinggroups/serverless.md:
```diff
@@ -0,0 +1,39 @@
+
+## Objective
+
+Explore the intersection of cloud native and serverless technology.
```
As above, "explore" is an activity, not an objective. Could we please craft an actual objective?

bgrant0607 replied:
Or eliminate the object subsections, if redundant with goals.

bgrant0607 commented on workinggroups/serverless.md:
```diff
@@ -0,0 +1,39 @@
+
+## Goals and Expected Outcomes
+
+* Produce a whitepaper
```
Done, correct?

bgrant0607 commented on workinggroups/storage.md:
```diff
@@ -0,0 +1,32 @@
+
+## Objective
+
+Explore cloud native storage technology and concepts.
```
Same comment on "explore".

But we've also had discussions about what "storage" should encompass. I think the scope needs to be clarified and narrowed. We also need clarification on what the group will produce.

cc @benh @monadic @skamille

bgrant0607 commented on workinggroups/storage.md:
```diff
@@ -0,0 +1,32 @@
+## Goals and Expected Outcomes
+
+* Produce a landscape
+* Explore specifications for storage to propose to the CNCF
```
AFAIK, no specifications are within scope of this WG.

bgrant0607 commented on workinggroups/storage.md:
```diff
@@ -0,0 +1,32 @@
+
+* Produce a landscape
+* Explore specifications for storage to propose to the CNCF
+* Bring recommendations to the TOC on storage projects in CNCF
```
I would like to eliminate this as a goal. WGs are not gatekeepers for which projects the TOC considers.

ultrasaurus replied:
I didn't read this as implying that no one else could propose a storage project to the TOC.  Maybe could be re-worded as something like:

Evaluate storage project proposals, as well as potential outreach to projects that would be good CNCF projects, then provide data to the TOC on proposals

that's a little wordy, but I was trying to expand upon something discussed in a TOC meeting a while ago, where someone said that WG were supposed to help the TOC, providing more bandwidth with structure for volunteers to get involved, essentially expanding what the TOC can accomplish

caniszczyk commented:
"Bring recommendations to the TOC on storage projects in CNCF"

I don't think this is gatekeeping, I think this is allowing a group of
experts in a space to discuss projects and approaches. This doesn't imply
that a WG has to approve a project before the TOC considers it.

On Fri, Apr 20, 2018 at 2:44 PM, Brian Grant <notifications@github.com>
wrote:

> *@bgrant0607* commented on this pull request.
>
> We need to revisit what we're trying to achieve with WGs.
> ------------------------------
>
> In workinggroups/networking.md
> <https://github.com/cncf/toc/pull/106#discussion_r183149679>:
>
> > @@ -0,0 +1,53 @@
> +# CNCF Networking WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore cloud native networking technology and concepts around the container networking interface (CNI).
> +
> +## Goals and Expected Outcomes
> +
> +* Recommend CNI be adopted as initial network interface specification focused on connectivity and portability as an official CNCF project.
>
> I assume this is done. Otherwise, it should be under the CNI project
> rather than in a WG.
> ------------------------------
>
> In workinggroups/networking.md
> <https://github.com/cncf/toc/pull/106#discussion_r183149793>:
>
> > @@ -0,0 +1,53 @@
> +# CNCF Networking WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore cloud native networking technology and concepts around the container networking interface (CNI).
> +
> +## Goals and Expected Outcomes
>
> Does this WG still need to exist, or can it declare victory and dissolve?
> ------------------------------
>
> In workinggroups/networking.md
> <https://github.com/cncf/toc/pull/106#discussion_r183149923>:
>
> > @@ -0,0 +1,53 @@
> +# CNCF Networking WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore cloud native networking technology and concepts around the container networking interface (CNI).
>
> "Explore" seems like an activity rather than an objective. Is there an
> objective at this point?
> ------------------------------
>
> In workinggroups/serverless.md
> <https://github.com/cncf/toc/pull/106#discussion_r183150124>:
>
> > @@ -0,0 +1,39 @@
> +# CNCF Serverless WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore the intersection of cloud native and serverless technology.
>
> As above, "explore" is an activity, not an objective. Could we please
> craft an actual objective?
> ------------------------------
>
> In workinggroups/serverless.md
> <https://github.com/cncf/toc/pull/106#discussion_r183150203>:
>
> > @@ -0,0 +1,39 @@
> +# CNCF Serverless WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore the intersection of cloud native and serverless technology.
>
> Or eliminate the object subsections, if redundant with goals.
> ------------------------------
>
> In workinggroups/serverless.md
> <https://github.com/cncf/toc/pull/106#discussion_r183150344>:
>
> > @@ -0,0 +1,39 @@
> +# CNCF Serverless WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore the intersection of cloud native and serverless technology.
> +
> +## Goals and Expected Outcomes
> +
> +* Produce a whitepaper
> +* Produce a serverless landscape
> +* Explore specifications for serverless to propose to the CNCF
>
> s/Explore/Develop/
>
> "Explore" isn't a goal.
> ------------------------------
>
> In workinggroups/serverless.md
> <https://github.com/cncf/toc/pull/106#discussion_r183150433>:
>
> > @@ -0,0 +1,39 @@
> +# CNCF Serverless WG Proposal
> +
> +## TOC Sponsor
> +
> +Ken Owens
> +
> +## Objective
> +
> +Explore the intersection of cloud native and serverless technology.
> +
> +## Goals and Expected Outcomes
> +
> +* Produce a whitepaper
>
> Done, correct?
> ------------------------------
>
> In workinggroups/storage.md
> <https://github.com/cncf/toc/pull/106#discussion_r183151327>:
>
> > @@ -0,0 +1,32 @@
> +# CNCF Storage WG Proposal
> +
> +## TOC Sponsor
> +
> +Ben Hindman
> +
> +## Objective
> +
> +Explore cloud native storage technology and concepts.
>
> Same comment on "explore".
>
> But we've also had discussions about what "storage" should encompass. I
> think the scope needs to be clarified and narrowed. We also need
> clarification on what the group will produce.
>
> cc @benh <https://github.com/benh> @monadic <https://github.com/monadic>
> @skamille <https://github.com/skamille>
> ------------------------------
>
> In workinggroups/storage.md
> <https://github.com/cncf/toc/pull/106#discussion_r183151391>:
>
> > @@ -0,0 +1,32 @@
> +# CNCF Storage WG Proposal
> +
> +## TOC Sponsor
> +
> +Ben Hindman
> +
> +## Objective
> +
> +Explore cloud native storage technology and concepts.
> +
> +## Goals and Expected Outcomes
> +
> +* Produce a landscape
> +* Explore specifications for storage to propose to the CNCF
>
> AFAIK, no specifications are within scope of this WG.
> ------------------------------
>
> In workinggroups/storage.md
> <https://github.com/cncf/toc/pull/106#discussion_r183151515>:
>
> > @@ -0,0 +1,32 @@
> +# CNCF Storage WG Proposal
> +
> +## TOC Sponsor
> +
> +Ben Hindman
> +
> +## Objective
> +
> +Explore cloud native storage technology and concepts.
> +
> +## Goals and Expected Outcomes
> +
> +* Produce a landscape
> +* Explore specifications for storage to propose to the CNCF
> +* Bring recommendations to the TOC on storage projects in CNCF
>
> I would like to eliminate this as a goal. WGs are not gatekeepers for
> which projects the TOC considers.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cncf/toc/pull/106#pullrequestreview-114100798>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAD5IZYPovrjHQLlm_SActUpevMbRO4Tks5tqjqRgaJpZM4TENpW>
> .
>



-- 
Cheers,

Chris Aniszczyk
http://aniszczyk.org
+1 512 961 6719

caniszczyk commented:
It should be explicit, they should present their proposal at a TOC meeting
for approval IMHO.

A landscape or whitepaper coming from CNCF is a big thing and requires TOC
oversight IMHO

On Tue, Apr 24, 2018 at 6:26 PM, Matt Farina <notifications@github.com>
wrote:

> *@mattfarina* commented on this pull request.
> ------------------------------
>
> In workinggroups/README.md
> <https://github.com/cncf/toc/pull/106#discussion_r183835041>:
>
> > @@ -0,0 +1,11 @@
> +# CNCF Working Groups
> +
> +## Introduction
> +
> +The purpose of working groups are to study and report on a particular question and make recommendations based on its findings. The end result of a working group may be a new project proposal, landscape, whitepaper or even a report detailing their findings. The intention of working groups is not to host a full project or specification. Working Groups can be formed at any time but must be sponsored by a TOC member and voted with a super majority vote by the CNCF TOC. The TOC can also shut down a working group with a super majority vote.
>
> And if they publish information under the CNCF, such as landscapes or
> whitepapers, I think it makes sense to at least send them to the TOC for
> review, possibly with an explicit time bound for feedback ("RFC" period).
>
> Should it be implicit or explicit? For example, is lazy-concensus ok?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cncf/toc/pull/106#discussion_r183835041>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAD5IScRYj2TXffoNzMFgO3iZFk3gyTtks5tr25WgaJpZM4TENpW>
> .
>



-- 
Cheers,

Chris Aniszczyk
http://aniszczyk.org
+1 512 961 6719

mattfarina commented on workinggroups/ci.md:
```diff
@@ -0,0 +1,34 @@
+# CNCF CI WG Proposal
```
Could a template file be created with the sections that are wanted and comments (using HTML comments) to explain how to fill them in? It could be called `TEMPLATE.md` and everyone could use that as a starting point. It would lead them to fill in the details the way they are needed.

mattfarina replied:
I wrote up an idea for a template... https://gist.githubusercontent.com/mattfarina/5738bab306e3907c5c523156dbc21469/raw/6c6b9bfc42451ed2a78f3448e3632b250ab93e86/TEMPLATE.md

ultrasaurus replied:
I like the template!

ultrasaurus replied:
If this is a real proposal for a CI WG, then I think it should be split into separate PR.

mattfarina commented on workinggroups/README.md:
```diff
@@ -0,0 +1,11 @@
+## Process
+
+If you would like to submit a working group proposal, please submit a pull request to the working groups folder. As an example, you can see the other working group proposals here: https://github.com/cncf/toc/tree/master/workinggroups
+
```
Could it be noted that the PR's should include:

* Goals
* Non-goals
* Mailing list information
* The location of meetings / agenda / notes
* Initial interested parties to show that there are multiple people across multiple orgs interested
* The chair(s) being explicitly listed so they are discoverable

It would also be nice to note that the goals should be something that can be accomplished and that once they are complete the WG needs to revise the goals and have the revisions approved by the TOC or come to a close.

caniszczyk commented:
This has been open for awhile, merging it in after addressing some feedback. We have a Security WG being proposed on the TOC call tomorrow so this is a way of testing the process:

https://github.com/cn-security/safe/issues/43

We can iterate in case we find any issues, thanks all.
